File: golang/trie.go
package main

import (
	"fmt"
)

type Trie struct {
	isEnd bool
	nodes map[rune]*Trie
}

func NewTrie() *Trie {
	return &Trie{nodes: make(map[rune]*Trie)}
}

func (t *Trie) Insert(word string) {
	current := t
	for _, c := range word {
		if current.nodes[c] == nil {
			current.nodes[c] = NewTrie()
		}
		current = current.nodes[c]
	}
	current.isEnd = true
}

func (t *Trie) Search(word string) bool {
	current := t
	for _, c := range word {
		if current.nodes[c] == nil {
			return false
		}
		current = current.nodes[c]
	}
	return current.isEnd
}

func main() {
	t := NewTrie()
	t.Insert("hello")
	t.Insert("world")
	fmt.Println(t.Search("hello")) // Output: true
	fmt.Println(t.Search("world")) // Output: true
	fmt.Println(t.Search("hi")) // Output: false
}

File: rust/trie.rs
use std::collections::HashMap;

struct Trie {
    is_end: bool,
    nodes: HashMap<char, Trie>,
}

impl Trie {
    fn new() -> Trie {
        Trie {
            is_end: false,
            nodes: HashMap::new(),
        }
    }

    fn insert(&mut self, word: &str) {
        let mut current = self;
        for c in word.chars() {
            current = current.nodes.entry(c).or_insert(Trie::new());
        }
        current.is_end = true;
    }

    fn search(&self, word: &str) -> bool {
        let mut current = self;
        for c in word.chars() {
            if let Some(node) = current.nodes.get(&c) {
                current = node;
            } else {
                return false;
            }
        }
        current.is_end
    }
}

fn main() {
    let mut trie = Trie::new();
    trie.insert("hello");
    trie.insert("world");
    println!("{}", trie.search("hello")); // Output: true
    println!("{}", trie.search("world")); // Output: true
    println!("{}", trie.search("hi")); // Output: false
}
